internal/driver/common/canvas.go

```diff
@@ -114,6 +118,64 @@ func (c *Canvas) EnsureMinSize() bool {
 	return windowNeedsMinSizeUpdate
 }
 
+func (c *Canvas) ensureMinSize2() bool {
+	if c.impl.Content() == nil {
+		return false
+	}
+	var lastParent fyne.CanvasObject
+
+	seenObj := map[fyne.CanvasObject]struct{}{}
+
+	windowNeedsMinSizeUpdate := false
+	csize := c.impl.Size()
+	min := c.impl.MinSize()
+
+	ensureMinSize := func(node *RenderCacheNode) {
+		obj := node.obj
+		cache.SetCanvasForObject(obj, c.impl)
+
+		if !obj.Visible() {
+			return
+		}
+		minSize := obj.MinSize()
+		minSizeChanged := node.minSize != minSize
+		if minSizeChanged {
+			objToLayout := obj
+			node.minSize = minSize
+			if node.parent != nil {
+				objToLayout = node.parent.obj
+			} else {
+				windowNeedsMinSizeUpdate = true
+				size := obj.Size()
+				expectedSize := minSize.Max(size)
+				if expectedSize != size && size != csize {
+					objToLayout = nil
+					obj.Resize(expectedSize)
+				}
+			}
+
+			if _, hasSeen := seenObj[objToLayout]; !hasSeen {
+				updateLayout(objToLayout)
+				seenObj[objToLayout] = struct{}{}
+				lastParent = objToLayout
+			}
+		}
+	}
+	c.WalkTrees(nil, ensureMinSize)
+
+	shouldResize := windowNeedsMinSizeUpdate && (csize.Width < min.Width || csize.Height < min.Height)
+	if shouldResize {
+		c.impl.Resize(csize.Max(min))
+	}
+
+	if lastParent != nil {
+		c.RLock()
+		updateLayout(lastParent)
+		c.RUnlock()
+	}
+	return windowNeedsMinSizeUpdate
+}
```

Comment:
It's better to just add this in the original function instead of having both to benchmark in this way. I would suggest benchmarking before and after instead, using the git history to your help (i.e. benchmark on a different branch or by temporary going back to before the change).